File: 94. Binary Tree Inorder Traversal/94. Binary Tree Inorder Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        # Recursive solution
        def inorder_helper(node, result):
            if not node:
                return
            inorder_helper(node.left, result)
            result.append(node.val)
            inorder_helper(node.right, result)
        
        result = []
        inorder_helper(root, result)
        return result
    
    def inorderTraversalIterative(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        # Iterative solution
        result = []
        stack = []
        current = root
        
        while current or stack:
            while current:
                stack.append(current)
                current = current.left
            
            current = stack.pop()
            result.append(current.val)
            current = current.right
        
        return result

File: 94. Binary Tree Inorder Traversal/94. Binary Tree Inorder Traversal.js
function inorderTraversal(root) {
    const result = [];
    const stack = [];
    let current = root;

    while (current !== null || stack.length > 0) {
        // Reach the leftmost node of the current node
        while (current !== null) {
            stack.push(current);
            current = current.left;
        }

        // Current must be null at this point
        current = stack.pop();
        result.push(current.val); // Add the node value

        // We have visited the node and its left subtree, now it's time to visit the right subtree
        current = current.right;
    }

    return result;
}

// Definition for a binary tree node
function TreeNode(val, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
}

// Example Usage:
// Input: [1, null, 2, 3]
// Output: [1, 3, 2]
const root = new TreeNode(1, null, new TreeNode(2, new TreeNode(3)));
console.log(inorderTraversal(root)); // Output: [1, 3, 2]